portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java

```diff
@@ -341,24 +353,136 @@ public String getPrimaryDomainName() throws UserPortalUIException {
+                throw new UserPortalUIException(error);
+            }
+        }
+
+        userList = generateUserListBean(users, requestedClaims);
+
+        return userList;
+    }
+
+    @Override
+    public List<UserListBean> getUserList(int offset, int length, String domainName,
+                                          List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<User> users;
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+        try {
+            users = getRealmService().getIdentityStore().listUsers(offset, length, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while retrieving users";
+            LOGGER.error(error, e);
+            throw new UserPortalUIException(error);
+        }
+
+        return generateUserListBean(users, requestedClaims);
+    }
 
     private RealmService getRealmService() {
         if (this.realmService == null) {
             throw new IllegalStateException("Realm Service is null.");
         }
         return this.realmService;
     }
+
+    private List<UserListBean> generateUserListBean(List<User> users, List<ClaimConfigEntry> requestedClaims)
+            throws UserPortalUIException {
+        List<UserListBean> userList = new ArrayList<>();
+
+        Map<String, String> claimDisplayNames = requestedClaims.stream()
+                        .collect(Collectors.toMap(ClaimConfigEntry::getClaimURI, ClaimConfigEntry::getDisplayName));
+
+        List<MetaClaim> metaClaims = requestedClaims.stream()
+                .map(claimConfigEntry -> new MetaClaim("", claimConfigEntry.getClaimURI()))
+                .collect(Collectors.toList());
+
+
+        List<MetaClaim> groupMetaClaims = new ArrayList<>();
+        MetaClaim groupMetaClaim = new MetaClaim();
+        groupMetaClaim.setClaimUri(GROUPNAME_CLAIM);
+        groupMetaClaims.add(groupMetaClaim);
+
+        for (User user : users) {
+            List<Group> groups;
+            List<Claim> userClaims;
+            List<Claim> groupId;
+            List<String> groupNames = new ArrayList<>();
+            try {
+                groups = user.getGroups();
+                for (Group group : groups) {
+                    groupId = group.getClaims(groupMetaClaims);
+                    if (!groupId.isEmpty()) {
+                        groupNames.add(groupId.get(0).getValue());
+                    }
+                }
+                userClaims = user.getClaims(metaClaims);
+
+            } catch (IdentityStoreException | GroupNotFoundException | UserNotFoundException e) {
+                String error = "Error while retrieving user data for user :  " + user.getUniqueUserId();
+                LOGGER.error(error, e);
+                throw new UserPortalUIException(error);
+            }
+
+            Map<String, String> userClaimMap = new LinkedHashMap<>();
+            for (Claim claim : userClaims) {
```

Comment:
@omindu It's critical to use LinkedHashMap in this case. Can we enforce it with the lambda. With this suggested code, it only gives a normal HashMap